R/import_ebird.R

```diff
@@ -83,5 +150,56 @@ ebird_col_type <- function(col_names) {
   setNames(col_types, col_names)
 }
 
-# Download URLs, eg https://download.ebird.org/ebd/prepackaged/ebd_relJul-2021.tar
-# Note, may take > 24hrs to complete
+parse_ebd_filename <- function(tarfile) {
+  stopifnot(is.character(tarfile), length(tarfile) == 1, file.exists(tarfile))
+  
+  f <- basename(tarfile)
+  # checks for validity of filename
+  if (!grepl("\\.tar$", f)) {
+    stop("The provided file does not appear to be a tar archive. The file ",
+         "extension should be .tar.")
+  }
+  if (grepl("ebd_sampling_rel[A-Z]{1}[a-z]{2}-[0-9]{4}\\.tar$", f)) {
+    dataset <- "checklists"
+    subset <- NA_character_
+  } else if (grepl("ebd_rel[A-Z]{1}[a-z]{2}-[0-9]{4}\\.tar$", f)) {
+    dataset <- "observations"
+    subset <- NA_character_
+  } else if (grepl("ebd_[-_A-Za-z0-9]+_rel[A-Z]{1}[a-z]{2}-[0-9]{4}\\.tar$", f)) {
+    dataset <- "observations"
+    subset <- sub("ebd_([-_A-Za-z0-9]+)_rel[A-Z]{1}[a-z]{2}-[0-9]{4}\\.tar", 
+                  "\\1", f)
+  } else {
+    stop("The provided tar filename does not appear to contain eBird data. ", 
+         "The expected format is, e.g., ebd_relJul-2021.tar.")
+  }
+  
+  # parse date from filename
+  rawdate <- sub("ebd[-_A-Za-z0-9]*_rel([A-Z]{1}[a-z]{2}-[0-9]{4})\\.tar", 
+                 "\\1", f)
+  date <- strsplit(rawdate, "-")[[1]]
+  date[1] <- match(date[1], month.abb)
+  date <- paste(date[2], date[1], "1", sep = "-")
+  date <- as.Date(date, format = "%Y-%m-%d")
+  if (is.na(date)) {
+    stop("Month and year could not be parsed from filename: ", rawdate)
+  }
+  version = format(date, "%Y-%m")
+
+  msg <- sprintf("Importing %s data from the %s eBird Basic Dataset: %s",
+                 dataset, version, f)
+  message(msg)
+  if (!is.na(subset)) {
+    message("EBD subset detected for: ", subset)
+  }
+  
+  # sha256 file hash
+  hash <- openssl::sha256(file(tarfile))
```

Comment:
You mentioned adding a hash of the file. I've never done that before and came across this method through googling. Happy to change if you know a better way, e.g. is there a base R package that does this?

Reply:
Nice, this is good.  

There's no base R function. The same utilities can be found in the `digest` package, but the ones in `openssl` are better (and often much faster, especially on more modern CPUs which have explicit hardware acceleration for sha256 built into the chip!).